### PR TITLE
Feature/cc 729

### PR DESCRIPTION
**Pull request by tgjonestx**
First, cherry-pick fixes from develop for PR #1466 and then add additional locking for `restLogout()`

Also includes a fix for a deadlock in #1466 which can occur if no one logins for more than 1 minute after serviced starts. In that scenario, the previous impl of `purgeOldsessionTs()` would grab the lock and not release it

### CODE REVIEW

File: web/session.go [hunk 1]
```diff
@@ -83,6 +85,14 @@ func purgeOldsessionTs() {
 			delete(sessions, key)
 		}
 	}
+
+	for {
+		time.Sleep(time.Second * 60)
+
+		sessionsLock.Lock()
```
Comment (iancmcc) [inline web/session.go]: I would suggest you put the Lock() and then a defer Unlock() inside doPurge itself.

File: web/session.go [hunk 2]
```diff
@@ -112,7 +125,9 @@ func restLogout(w *rest.ResponseWriter, r *rest.Request) {
 	if err != nil {
 		glog.V(1).Info("Unable to read session cookie")
 	} else {
+		sessionsLock.Lock()
```
Comment (iancmcc) [inline web/session.go]: Ditto. Wrap this in a function so you can use defer.

Comment (iancmcc) [inline web/session.go]: Or actually just use defer.

File: web/session.go (hunk 1 above)
Comment (tgjonestx) [inline web/session.go]: I found a bug in the original impl where it failed to Unlock if there were 0 sessions. So if you start, wait > 1 minute, then you can't login. So I could have fixed that bug with an Unlock call before the continue, but I thought that left the routine vulnerable to some future bug if another continue was added.  I can move the Lock/Unlock into the closure if you think that's better.

Comment (iancmcc) [inline web/session.go]: Yeah, I'd do something like:

if len(sessions) == 0 {
    return
}
sessionsLock.Lock()
defer sessionsLock.Unlock()

**Pull request merged**